---
title: "session 2: NGS Sequencing Technology and File Formats"
author: "Archibald Worwui, Mouhamadou F. DIOP, Lucas A. Etego & Alfred A. NGWA"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
    powerpoint_presentation:
        reference_doc: MRC_Slides_Template.pptx
        slide_level: 3
        
---

## NGS Sequencing Technology and File Formats 

There are a variety different sequencing technologies as well as file formats used in sequence analysis.

- Describe how next-generation sequencing works 
- Most common File formats including those generated by the sequencer and analysis programs.

## Definitions

- Read: a single sequence produced from a sequencer. 

- Library: a collection of DNA fragments that have been prepared for sequencing.

- Flowcell: a chip on which DNA is loaded and provided to the sequencer.

- Lane: one portion of a flowcell. Usually used for technical replicates or different samples.

- Run: an entire sequencing reaction from start to finish.

## FastA Format

The official FastA documentation can be found [here](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp)

Most basic format for reporting a sequence and is accepted by almost all sequence analysis program. 
Contains only:
    Sequence name, 
    Description of the sequence (metadata, sequencer info, annotations, etc.), 
    Sequence itself
    
it can be either nucleic acids or amino acids as long as it adheres to the format.

---

Each sequence consists of at least two lines:

1. The first is the sequence header, which always starts with a **‘>’**

* Everything from the beginning ‘>’ to the first whitespace is considered the sequence identifier. 
* Everything after that is considered the sequence description (this can be metadata, machine serial number, read orientation, etc.)

--- 
2. The sequence itself

* Note that the sequence can span multiple lines, depending on the length of the sequence.

```{bash echo = TRUE, eval=FALSE}
>Chr1 CHROMOSOME dumped from ADB: Jun/20/09 14:53; last updated: 2009-02-02
CCCTAAACCCTAAACCCTAAACCCTAAACCTCTGAATCCTTAATCCCTAAATCCCTAAATCTTTAAATCCTACATCCAT
GAATCCCTAAATACCTAATTCCCTAAACCCGAAACCGGTTTCTCTGGTTGAAAATCATTGTGTATATAATGATAATTTT
ATCGTTTTTATGTAATTGCTTATTGTTGTGTGTAGATTTTTTAAAAATATCATTTGAGGTCAATACAAATCCTATTTCT
TGTGGTTTTCTTTCCTTCACTTAGCTATGGATGGTTTATCTTCATTTGTTATATTGGATACAAGCTTTGCTACGATCTA
CATTTGGGAATGTGAGTCTCTTATTGTAACCTTAGGGTTGGTTTATCTCAAGAATCTTATTAATTGTTTGGACTGTTTA
TGTTTGGACATTTATTGTCATTCTTACTCCTTTGTGGAAATGTTTGTTCTATCAATTTATCTTTTGTGGGAAAATTATT
TAGTTGTAGGGATGAAGTCTTTCTTCGTTGTTGTTACGCTTGTCATCTCATCTCTCAATGATATGGGATGGTCCTTTAG
```

## Software that use FastA format

In most case throughout this workshop you will encounter this format when using a reference sequence. 
- [blast](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch)
- Multiple-sequence alignment algorithms. 

Also, when you download reference genomes they are delivered in this format.

## How are these files generated?
* Some older NGS sequencers report sequences in this format. Sanger sequencing also delivers in this format.
* Most sequence databases store sequences in FastA format which is available for download.
* FastA can also generated from a FastQ file.

## FastQ Format
The official documentation for FastQ format can be found [here](https://maq.sourceforge.net/fastq.shtml).

- Most widely used format in sequence analysis
- Generally delivered from a sequencer. 
- Many analysis tools require this format.

Similar to fasta though there are differences in syntax as well as integration of quality scores.
Each sequence requires at least 4 lines:

1. The first line is the sequence header which starts with an ‘@’ (not a ‘>’!).
    * Everything from the leading ‘@’ to the first whitespace character is considered the sequence identifier.
    * Everything after the first space is considered the sequence description
2. The second line is the sequence.
3. The third line starts with ‘+’ and can have the same sequence identifier appended (but usually doesn’t anymore).
4. The fourth line are the quality scores

The FastQ sequence identifier generally adheres to a particular format, all of which is information related to the sequencer and its position on the flowcell. The sequence description also follows a particular format and holds information regarding sample information.

![](images/fastq.png)<!-- -->

## What software use FastQ?
Nearly everything works with this format. Some common examples are:

- Aligners
    * Bowtie, Tophat2
- Assemblers
    * Velvet, Spades
- QC tools
    * Trimmomatic, FastQC

I think it’s a shorter list to tell you what does not work with FastQ files. Please note that there are tools available to convert FastQ to FastA in the event that FastQ is incompatible with the tool you’re using:

- Blast
- Multiple Sequence Aligners
- Any reference sequence

## How are these files generated?
- Sequencers generate this format by default.
- This can also be generated from a few different file formats (BAM, SFF, HDF5), though they all were some form of FastQ at some point.

## Quality Scores
An in depth writeup about quality scores can be found [here](https://www.illumina.com/documents/products/technotes/technote_Q-Scores.pdf).

Quality scores are a way to assign confidence to a particular base within a read. Some sequencers have their own proprietary quality encoding but most have adopted Phred-33 encoding. Each quality score represents the probability of an incorrect basecall at that position.

### Phred Quality Score Encoding
Quality scores started as numbers (0-40) but have since changed to an ASCII encoding to reduce file size and make working with this format a bit easier, however they still hold the same information. ASCII codes are assigned based on the formula found below. This table can serve as a lookup as you progress through your analysis.

**Note that Phred-64 was only ever used by Illumina and is not deprecated.**

IMAGE: phred score

## Quality Score Interpretation
Once you know what each quality score represents you can then use this chart to understand the confidence in a particular base.

image: quality interpretation

## What are Quality Scores Good for?
As we mentioned earlier, many programs require the FastQ format, implying that they will use the quality score in a particular part of the analysis. Common uses are to filter bases or entire reads if a particular quality threshold isn’t met. An example of a threshold is the mean quality score for the read. That is: what’s the average score of all bases for an individual read? If the average Phred quality score is 10, what does that mean? Is this good enough to do SNP analysis?

image

## What Software use Quality Scores?
Provide evidence that the sequence, alignment, assembly, SNP are in fact real and not due to a problem in generating the sequences.

- Almost every QC software package use these.
- Variant detection/SNP calling algorithms
- Assemblers
- Aligners

## SAM/BAM/CRAM Format

The official SAM documentation can be found [here](https://samtools.github.io/hts-specs/SAMv1.pdf).

These formats were introduced to standardize how alignments are reported. Initially there were many different formats, most of them proprietary, which were space inefficient and either held too much or too little information. The first of these to be introduced was Sequence Alignment Map (SAM). With this format not only is the alignment retained but the associated quality scores (both mapping and base quality), the original read itself, paired-end information, sample information, and many more features.

## SAM Format
Most basic, human readable format of the three. 
Generated by almost every alignment algorithm that exists. 
Consists of a header, a row for every read
11 tab-delimited fields describing that read.

## SAM Header
The header varies in size but adheres to a particular format depending on what information you decide to add. Some example information that can be entered into the header is: command that generated the SAM file, SAM format version, sequencer name and version. The full list of available header fields can be found below.

## BAM Format
This is the same format except that it encoded in binary which means that it is significantly smaller than the SAM files and significantly faster to read, though it is not human legible and needs to be converted to another format (i.e. SAM) in order to make sense to us.

Some special tools are needed in order to make sense of BAM, such as [Samtools](https://www.htslib.org/), [Picard Tools](https://broadinstitute.github.io/picard/), and [IGV](https://software.broadinstitute.org/software/igv/) which will be discussed in some of the latter sections.

## CRAM Format
This is a relatively new format that is very similar to BAM as it also retains the same information as SAM and is compressed, but it is much smarter in the way that it stores the information. It’s very interesting and up and coming but is a bit beyond the scope of this course. However, if you’re up for it you can read about it [here](https://samtools.github.io/hts-specs/CRAMv3.pdf).

## What software use these files?
- Alignment algorithms
- Some assemblers
- CRAM/unaligned Bam (uBAM) can be a source of data delivery in some institutions: this cuts down significantly on storage space and transfer speed.
- Alignment viewers
- Variant detection algorithms

## How are these files generated?
This is output from aligners and assemblers
This can also be used to deliver raw data

## BED Format
The official documentation for BED format can be found [here](https://genome.ucsc.edu/FAQ/FAQformat.html#format1).

BED format is a simple way to define basic sequence features to a sequence. It consists of one line per feature, each containing 3-12 columns of data, plus optional track definition lines. These are generally used for user defined sequence features as well as graphical representations of features.

Here are some links to the formal definitions of each field.

- Required fields
- Optional fields
- Track lines
- BedGraph format

## Required fields
The first three fields in each feature line are required:

1. Chromosome Name
    * Name of the chromosome or scaffold. Any valid seq\_region\_name can be used, and chromosome names can be given with or without the ‘chr’ prefix.
2. Chromosome Start
    * Start position of the feature in standard chromosomal coordinates \(i.e. first base is 0\).
3. Chromosome End
    * End position of the feature in standard chromosomal coordinates
    
```{bash echo = TRUE, eval=FALSE}
chr1 213941196 213942363
chr1 213942363 213943530
chr1 213943530 213944697
chr2 158364697 158365864
chr2 158365864 158367031
chr3 127477031 127478198
chr3 127478198 127479365
chr3 127479365 127480532
chr3 127480532 127481699
```

## Optional fields
Nine additional fields are optional. Note that columns cannot be empty – lower-numbered fields must always be populated if higher-numbered ones are used.

1. Name
    Label to be displayed under the feature, if turned on in “Configure this page”.
2. Score
    A score between 0 and 1000. 
    See [track lines](http://asia.ensembl.org/info/website/upload/bed.html#tracklines)
, below, for ways to configure the display style of scored data.
3. Strand
    defined as + \(forward\) or – \(reverse\).
4. thickStart
    coordinate at which to start drawing the feature as a solid rectangle
5. thickEnd
c       oordinate at which to stop drawing the feature as a solid rectangle
6. itemRgb
    an RGB colour value \(e.g. 0,0,255\). Only used if there is a track line with the value of itemRgb set to “on” \(case-insensitive\).
7. blockCount
    the number of sub-elements \(e.g. exons\) within the feature
8. blockSizes
    the size of these sub-elements
9. blockStarts
    the start coordinate of each sub-element

```{bash echo = TRUE, eval=FALSE}
chr7 127471196 127472363 Pos1 0 + 127471196 127472363 255,0,0
chr7 127472363 127473530 Pos2 0 + 127472363 127473530 255,0,0
chr7 127473530 127474697 Pos3 0 + 127473530 127474697 255,0,0
chr7 127474697 127475864 Pos4 0 + 127474697 127475864 255,0,0
chr7 127475864 127477031 Neg1 0 - 127475864 127477031 0,0,255
chr7 127477031 127478198 Neg2 0 - 127477031 127478198 0,0,255
chr7 127478198 127479365 Neg3 0 - 127478198 127479365 0,0,255
chr7 127479365 127480532 Pos5 0 + 127479365 127480532 255,0,0
chr7 127480532 127481699 Neg4 0 - 127480532 127481699 0,0,255

```

## Track lines
Track definition lines can be used to configure the display further, e.g. by grouping features into separate tracks. Track lines should be placed at the beginning of the list of features they are to affect.

The track line consists of the word ‘track’ followed by space-separated key=value pairs – see the example below. Valid parameters used by Ensembl are:

name: unique name to identify this track when parsing the file
description: Label to be displayed under the track in Region in Detail
priority: integer defining the order in which to display tracks, if multiple tracks are defined.
color: as RGB, hex or X11 named color
useScore: a value from 1 to 4, which determines how scored data will be displayed. Additional parameters may be needed, as described below.
tiling array (example file)
colour gradient – defaults to Yellow-Green-Blue, with 20 colour grades. Optionally you can specify the colours for the gradient (cgColour1, cgColour2, cgColour3) as either RGB, hex or X11 colour names, and the number of colour grades (cgGrades). (example file)
histogram (example file)
wiggle plot (example file)
itemRgb
if set to ‘on’ (case-insensitive), the individual RGB values defined in tracks will be used.

```{bash echo = TRUE, eval=FALSE}
track name="ItemRGBDemo" description="Item RGB demonstration" itemRgb="On"
chr7 127471196 127472363 Pos1 0 + 127471196 127472363 255,0,0
chr7 127472363 127473530 Pos2 0 + 127472363 127473530 255,0,0
chr7 127473530 127474697 Pos3 0 + 127473530 127474697 255,0,0
chr7 127474697 127475864 Pos4 0 + 127474697 127475864 255,0,0
chr7 127475864 127477031 Neg1 0 - 127475864 127477031 0,0,255
chr7 127477031 127478198 Neg2 0 - 127477031 127478198 0,0,255
chr7 127478198 127479365 Neg3 0 - 127478198 127479365 0,0,255
chr7 127479365 127480532 Pos5 0 + 127479365 127480532 255,0,0
chr7 127480532 127481699 Neg4 0 - 127480532 127481699 0,0,255
```

## BedGraph format
BedGraph is a suitable format for moderate amounts of scored data. It is based on the BED format (see above) with the following differences:

The score is placed in column 4, not column 5
Track lines are compulsory, and must include type=bedGraph. Currently the only optional parameters supported by Ensembl are:
* name
* see above
* description
* see above
* priority
* see above
* graphType
* either ‘bar’ or ‘points’.

```{bash echo = TRUE, eval=FALSE}
track type=bedGraph name="BedGraph Format" description="BedGraph format" priority=20
chr19 59302000 59302300 -1.0
chr19 59302300 59302600 -0.75
chr19 59302600 59302900 -0.50
chr19 59302900 59303200 -0.25
chr19 59303200 59303500 0.0
chr19 59303500 59303800 0.25
chr19 59303800 59304100 0.50
chr19 59304100 59304400 0.75
```

## What software use bed files?

- Alignment viewers can use these data to graphically display certain features.
- [bedtools](http://bedtools.readthedocs.io/en/latest/index.html) uses this format to query for nearby features.
- Some annotation files are in this format.
- Feature detection packages use this as output.

## How are these files generated?
- Feature detection algorithms.
- Lots of databases that hold certain genomic features report their data in this format.
- Sometimes manually curated from alignments (via bedtools, bamtools, etc.).

## VCF Format

Variant Calling Format is a tab-delimited text file that is used to describe single nucleotide variants (SNVs) as well as insertions, deletions, and other sequence variations. This is a bit limiting as it is only tailored to show variations and not genetic features (that’ll be covered on the next page).

There are 8 required fields for this format:

1. Chromosome Name
2. Chromosome Position
3. ID
    - This is generally used to reference an annotated variant in dbSNP or other curate variant database.
4. Reference base(s)
    - What is the reference’s base at this position
5. Alternate base(s)
    - The variants found in your dataset that differ from the reference
6. Variant Quality
    - Phred-scaled quality for the observed ALT
7.Filter
    - Whether or not this has passed all filters – generally a QC measure in variant calling algorithms
8. Info
    - This is for additional information, generally describing the nature of the position/variants with respect to other data.
    
## Example VCF File
image

## What software use VCF?
- Output of SNP detection tools such as [GATK](https://software.broadinstitute.org/gatk/) and [Samtools](http://samtools.github.io/)
- Input for SNP feature detection like [SNPeff](http://snpeff.sourceforge.net/)
- [VCF Tools](https://vcftools.github.io/index.html)
- Also the required format for [dbSNP](https://www.ncbi.nlm.nih.gov/projects/SNP/)

## How are these files generated?
- SNP callers generate these files as output.
- Haplotyping software also report in this format.
- Any database holding variant information will generally have this format available for download.

## GFF3 Format
The official documentation for the GFF3 format can be found [here](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md)

General Feature Format (GFF) is a tab-delimited text file that holds information any and every feature that can be applied to a nucleic acid or protein sequence. Everything from CDS, microRNAs, binding domains, ORFs, and more can be handled by this format. Unfortunately there have been many variations of the original GFF format and many have since become incompatible with each other. The latest accepted format (GFF3) has attempted to address many of the issues that were missing from previous versions.

GFF3 has 9 required fields, though not all are utilized (either blank or a default value of ‘.’).

1. Sequence ID
2. Source
    - Describes the algorithm or the procedure that generated this feature. Typically Genescane or Genebank, respectively.
3. Feature Type
    - Describes what the feature is (mRNA, domain, exon, etc.).
    - These terms are constrained to the [Sequence Ontology terms](http://www.sequenceontology.org/).
4. Feature Start
5. Feature End
6. Score
    - Typically E-values for sequence similarity and P-values for predictions.
7. Strand
8. Phase
    - Indicates where the feature begins with reference to the reading frame. The phase is one of the integers 0, 1, or 2, indicating the number of bases that should be removed from the beginning of this feature to reach the first base of the next codon.
9. Atributes
    - A semicolon-separated list of tag-value pairs, providing additional information about each feature. Some of these tags are predefined, e.g. ID, Name, Alias, Parent .
    You can see the full list [here](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md).

## GFF3 Example
The canonical gene can be represented by the following figure

https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md

## The same information can be represented in GFF3 format:

https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md

## What Software uses GFF3?
-Any tool that requires information about gene position for analysis such as:
    * Mapping RNA-seq such as [Tophat](https://ccb.jhu.edu/software/tophat/index.shtml), [HTSeq](https://htseq.readthedocs.io/en/release_0.9.1/)
    * Genome Browsers like [IGV](http://software.broadinstitute.org/software/igv/), [Gbrowse](http://gmod.org/wiki/GBrowse), [UCSC](https://genome.ucsc.edu/)


## How is this file generated?
- Feature identification software report motifs/features in this format.
- Almost all sequence annotation databases report in this format.